In [31]:
import pandas as pd
import numpy as np
import scipy.stats as ss
import statsmodels.api as sm

First I am inputting the Data from the Portfolio tabs I created from the excel workshop data, and the new Indexes/YAS files obtained.

In [32]:
portfolio = pd.read_excel(io="Workshop Data.xlsx",sheet_name="Portfolio")
portfolio = portfolio.set_index("Date")
portfolio.head()

Total Value  Portf. Close  Cash of the Portfolio (Dividends)  \
Date                                                                        
2025-12-31  26486.550990  21019.672681                        5466.878309   
2025-12-30  26568.023255  21106.797740                        5461.225515   
2025-12-29  26570.820422  21115.247761                        5455.572661   
2025-12-26  26568.567583  21118.647757                        5449.919826   
2025-12-24  26625.572657  21180.897755                        5444.674902   

            Total Day CF  HYG Position  HYG Dividends  LQD Position  \
Date                                                                  
2025-12-31      5.652794   8061.187692       1.812511  22033.843693   
2025-12-30      5.652854   8069.187720       1.812484  22127.843614   
2025-12-29      5.652835   8061.187692       1.812511  22153.843655   
2025-12-26      5.244924   8058.187695       1.812508  22123.843663   
2025-12-24      5.248645   8062.187724       1.812480  22125.843639   

            LQD Dividends  SPY Position  SPY Dividends  TLT Position  \
Date                                                                   
2025-12-31       4.156699  -6892.483704      -0.316416     -2182.875   
2025-12-30       4.156776  -6895.483594      -0.316406     -2194.750   
2025-12-29       4.156735  -6917.783587      -0.316411     -2182.000   
2025-12-26       4.156727  -6876.883601      -0.316419     -2186.500   
2025-12-24       4.156752  -6831.383607      -0.316397     -2175.750   

            TLT Dividends  
Date                       
2025-12-31       0.000000  
2025-12-30       0.000000  
2025-12-29       0.000000  
2025-12-26      -0.407893  
2025-12-24      -0.404190

In [65]:
HYGVol = pd.read_excel(io="Workshop Data.xlsx",sheet_name="Adj HYG")
HYGVol = HYGVol.set_index("Date")
HYGVol = HYGVol.loc[:,"Volume"]
HYGVol.head()

Date
2026-01-08    22342728
2026-01-07    27022770
2026-01-06    31721460
2026-01-05    38563220
2026-01-02    28403304
Name: Volume, dtype: int64

In [33]:
LQDyas = pd.read_excel(io="Indexes and Spreads Data 01.09.xlsx",sheet_name="LQD")
LQDyas = LQDyas.set_index("Date")
LQDyas.head()

YAS_BOND_YLD  YAS_MOD_DUR  YAS_ISPREAD  YAS_YLD_SPREAD  YAS_RISK  \
Date                                                                           
2026-01-09        5.3109       8.3995     137.0123        114.4544    9.2932   
2026-01-08        5.3496       8.3941     140.1891        116.5661    9.2537   
2026-01-07        5.3150       8.4080     138.5779        117.4304    9.2950   
2026-01-06        5.3184       8.4233     136.2275        115.4190    9.3052   
2026-01-05        5.3244       8.4004     137.1802        116.6149    9.2799   

            YAS_ETF_ISPREAD_BASIS  YAS_ISPREAD_TO_GOVT  
Date                                                    
2026-01-09               -87.9011              91.5288  
2026-01-08               -90.9663              93.0004  
2026-01-07               -88.9592              93.6321  
2026-01-06               -87.1339              90.9014  
2026-01-05               -87.9253              92.5261

In [117]:
HYGyas = pd.read_excel(io="Indexes and Spreads Data 01.09.xlsx",sheet_name="HYG")
HYGyas = HYGyas.set_index("Date")
HYGyas = pd.DataFrame(HYGyas)
HYGyas.head()

YAS_BOND_YLD  YAS_MOD_DUR  YAS_ISPREAD  YAS_BOND_YLD.1  \
Date                                                                 
2026-01-09        6.7400       2.8204     335.9253          6.7400   
2026-01-08        6.7518       2.8146     340.6738          6.7518   
2026-01-07        6.7738       2.8006     345.2194          6.7738   
2026-01-06        6.7490       2.8030     342.5065          6.7490   
2026-01-05        6.8023       2.8196     348.7079          6.8023   

            YAS_YLD_SPREAD  YAS_ISPREAD_TO_GOVT  YAS_RISK  \
Date                                                        
2026-01-09        299.0917             311.6285    2.2846   
2026-01-08        301.1570             315.3019    2.2781   
2026-01-07        307.8833             321.3189    2.2651   
2026-01-06        304.5343             319.0898    2.2677   
2026-01-05        309.8603             324.2391    2.2805   

            YAS_ETF_ISPREAD_BASIS  
Date                               
2026-01-09               -23.2910  
2026-01-08               -27.3764  
2026-01-07               -29.5670  
2026-01-06               -29.8644  
2026-01-05               -35.5195

In [35]:
TenYearYield = pd.read_excel(io="Indexes and Spreads Data 01.09.xlsx",sheet_name="10yUST Yields")
TenYearYield = TenYearYield.set_index("Date")
TenYearYield.head()

PX_OPEN  PX_MID  PX_LAST  PX_LOW  PX_HIGH
Date                                                 
2026-01-09   4.1692  4.1653   4.1653  4.1575   4.2028
2026-01-08   4.1534  4.1672   4.1672  4.1299   4.1849
2026-01-07   4.1691  4.1475   4.1475  4.1221   4.1691
2026-01-06   4.1612  4.1730   4.1730  4.1593   4.1947
2026-01-05   4.1867  4.1612   4.1612  4.1455   4.1926

In [36]:
HYyield = pd.read_excel(io="Indexes and Spreads Data 01.09.xlsx",sheet_name="HY Index")
HYyield = HYyield.set_index("Date")
HYyield.head()

OAS_SOVEREIGN_CURVE  PX_LAST  INDEX_SPREAD_BENCHMARK  \
Date                                                               
2026-01-09             2.568904  2925.95              280.194871   
2026-01-08             2.580217  2923.30              282.848226   
2026-01-07             2.605467  2923.37              285.627406   
2026-01-06             2.591620  2921.87              284.360667   
2026-01-05             2.621012  2919.53              287.477672   

            INDEX_Z_SPREAD_BP  
Date                           
2026-01-09           281.1432  
2026-01-08           282.3889  
2026-01-07           284.7707  
2026-01-06           283.2562  
2026-01-05           286.6346

In [37]:
IGyield = pd.read_excel(io="Indexes and Spreads Data 01.09.xlsx",sheet_name="IG Index")
IGyield = IGyield.set_index("Date")
IGyield.head()

OAS_SOVEREIGN_CURVE  INDEX_SPREAD_BENCHMARK  PX_LAST  \
Date                                                               
2026-01-09             0.766109               77.571031  3548.84   
2026-01-08             0.775954               78.534368  3541.24   
2026-01-07             0.776639               78.540815  3549.91   
2026-01-06             0.766026               77.514646  3545.48   
2026-01-05             0.775938               78.517540  3544.78   

            INDEX_Z_SPREAD_BP  
Date                           
2026-01-09            77.4409  
2026-01-08            78.4368  
2026-01-07            78.5199  
2026-01-06            77.4352  
2026-01-05            78.4970

In [38]:
portfolio.loc[:,"HYGCumDiv"]=portfolio.loc[:,"HYG Dividends"][::-1].cumsum()[::-1]
portfolio.tail()

Total Value  Portf. Close  Cash of the Portfolio (Dividends)  \
Date                                                                        
2021-01-15  28269.066126  28249.419253                          19.646873   
2021-01-14  28296.987586  28281.269243                          15.718343   
2021-01-13  28411.709376  28399.919292                          11.790084   
2021-01-12  28148.529659  28140.669256                           7.860403   
2021-01-11  28131.749877  28127.819295                           3.930582   

            Total Day CF  HYG Position  HYG Dividends  LQD Position  \
Date                                                                  
2021-01-15      3.928530   8709.289908       1.710260  27225.333407   
2021-01-14      3.928259   8723.289894       1.710273  27183.333422   
2021-01-13      3.929680   8725.289918       1.710250  27301.333440   
2021-01-12      3.929822   8699.289904       1.710264  27091.333415   
2021-01-11      3.930582   8688.289928       1.710240  27033.333438   

            LQD Dividends  SPY Position  SPY Dividends  TLT Position  \
Date                                                                   
2021-01-15       2.666723  -3843.649209      -0.250795  -3841.554853   
2021-01-14       2.666709  -3828.549222      -0.250790  -3796.804851   
2021-01-13       2.666691  -3842.149215      -0.250797  -3784.554850   
2021-01-12       2.666716  -3838.649215      -0.250800  -3811.304847   
2021-01-11       2.666694  -3786.249213      -0.250792  -3807.554858   

            TLT Dividends  HYGCumDiv  
Date                                  
2021-01-15      -0.197659   8.551287  
2021-01-14      -0.197933   6.841027  
2021-01-13      -0.196463   5.130754  
2021-01-12      -0.196359   3.420505  
2021-01-11      -0.195560   1.710240

In [ ]:
portfolio.loc[:,"TotalReturnsHYG"]=portfolio.loc[:,"HYG Position"]+portfolio.loc[:,"HYGCumDiv"]
portfolio.tail()

Total Value  Portf. Close  Cash of the Portfolio (Dividends)  \
Date                                                                        
2021-01-15  28269.066126  28249.419253                          19.646873   
2021-01-14  28296.987586  28281.269243                          15.718343   
2021-01-13  28411.709376  28399.919292                          11.790084   
2021-01-12  28148.529659  28140.669256                           7.860403   
2021-01-11  28131.749877  28127.819295                           3.930582   

            Total Day CF  HYG Position  HYG Dividends  LQD Position  \
Date                                                                  
2021-01-15      3.928530   8709.289908       1.710260  27225.333407   
2021-01-14      3.928259   8723.289894       1.710273  27183.333422   
2021-01-13      3.929680   8725.289918       1.710250  27301.333440   
2021-01-12      3.929822   8699.289904       1.710264  27091.333415   
2021-01-11      3.930582   8688.289928       1.710240  27033.333438   

            LQD Dividends  SPY Position  SPY Dividends  TLT Position  \
Date                                                                   
2021-01-15       2.666723  -3843.649209      -0.250795  -3841.554853   
2021-01-14       2.666709  -3828.549222      -0.250790  -3796.804851   
2021-01-13       2.666691  -3842.149215      -0.250797  -3784.554850   
2021-01-12       2.666716  -3838.649215      -0.250800  -3811.304847   
2021-01-11       2.666694  -3786.249213      -0.250792  -3807.554858   

            TLT Dividends  HYGCumDiv  TotalReturnsHYG  
Date                                                   
2021-01-15      -0.197659   8.551287      8717.841195  
2021-01-14      -0.197933   6.841027      8730.130922  
2021-01-13      -0.196463   5.130754      8730.420672  
2021-01-12      -0.196359   3.420505      8702.710408  
2021-01-11      -0.195560   1.710240      8690.000168

Now, I am limiting the dates index so that all the data have the same number of entries and is consistent.

In [118]:
portfolio = portfolio.sort_index()
HYyield = HYyield.sort_index()
IGyield = IGyield.sort_index()
TenYearYield = TenYearYield.sort_index()
HYGVol = HYGVol.sort_index()
HYGyas = HYGyas.sort_index()
HYGCorrData = pd.DataFrame(portfolio.loc["2021-01-11":"2025-12-31","TotalReturnsHYG"]).merge(HYyield.loc["2021-01-11":"2025-12-31"],left_index=True,right_index=True)
HYGCorrData = HYGCorrData.rename(columns={"PX_LAST":"HY_INDEX_PX","OAS_SOVEREIGN_CURVE":"HY_Index_OAS","INDEX_SPREAD_BENCHMARK":"HY_INDEX_Spread","INDEX_Z_SPREAD_BP":"HY_INDEX_Z_Spread"})
HYGCorrData = HYGCorrData.loc["2021-01-11":"2025-12-31"].merge(TenYearYield.loc["2021-01-11":"2025-12-31","PX_LAST"],left_index=True,right_index=True)
HYGCorrData = HYGCorrData.rename(columns={"PX_LAST":"10y_UST_INDEX_PX"})
HYGCorrData = HYGCorrData.loc["2021-01-11":"2025-12-31"].merge(IGyield.loc["2021-01-11":"2025-12-31"],left_index=True,right_index=True)
HYGCorrData = HYGCorrData.rename(columns={"PX_LAST":"IG_INDEX_PX","OAS_SOVEREIGN_CURVE":"IG_Index_OAS","INDEX_SPREAD_BENCHMARK":"IG_INDEX_Spread","INDEX_Z_SPREAD_BP":"IG_INDEX_Z_Spread"})
HYGCorrData = HYGCorrData.loc["2021-01-11":"2025-12-31"].merge(HYGVol.loc["2021-01-11":"2025-12-31"],left_index=True,right_index=True)
HYGCorrData = HYGCorrData.rename(columns={"Volume":"HYG_Volume"})
HYGCorrData = HYGCorrData.loc["2021-01-11":"2025-12-31"].merge(portfolio.loc["2021-01-11":"2025-12-31","SPY Position"]/-10,left_index=True,right_index=True)
HYGCorrData = HYGCorrData.rename(columns={"SPY Position":"SPY_LAST_PX"})
HYGCorrData = HYGCorrData.loc["2021-01-11":"2025-12-31"].merge(HYGyas.loc["2021-01-11":"2025-12-31","YAS_MOD_DUR"],left_index=True,right_index=True)
HYGCorrData = HYGCorrData.rename(columns={"YAS_MOD_DUR":"HYG_MOD_DUR_YAS"})

HYGCorrData = HYGCorrData.sort_index(ascending=False)


HYGCorrData.head()

TotalReturnsHYG  HY_Index_OAS  HY_INDEX_PX  HY_INDEX_Spread  \
Date                                                                      
2025-12-31     10168.486998      2.663664      2914.49       291.500404   
2025-12-30     10174.674513      2.673836      2914.22       293.499232   
2025-12-29     10164.862002      2.711314      2911.40       297.866454   
2025-12-26     10160.049494      2.709540      2908.63       294.344862   
2025-12-24     10162.237014      2.671766      2909.19       291.571921   

            HY_INDEX_Z_Spread  10y_UST_INDEX_PX  IG_Index_OAS  \
Date                                                            
2025-12-31           293.7437            4.1670      0.775273   
2025-12-30           294.6537            4.1219      0.769639   
2025-12-29           298.9804            4.1102      0.772791   
2025-12-26           298.7869            4.1277      0.764817   
2025-12-24           294.5280            4.1335      0.768318   

            IG_INDEX_Spread  IG_INDEX_PX  IG_INDEX_Z_Spread  HYG_Volume  \
Date                                                                      
2025-12-31        78.399397      3545.18            78.6033    17457064   
2025-12-30        77.956202      3549.46            78.0161    20146744   
2025-12-29        78.925315      3551.49            78.3505    22012766   
2025-12-26        78.086383      3546.54            77.5529    10922908   
2025-12-24        78.476640      3545.84            77.9190    10533489   

            SPY_LAST_PX  HYG_MOD_DUR_YAS  
Date                                      
2025-12-31   689.248370           2.8197  
2025-12-30   689.548359           2.8482  
2025-12-29   691.778359           2.8465  
2025-12-26   687.688360           2.8596  
2025-12-24   683.138361           2.8713

**Correlation Table**

In [159]:
HYGCorr = HYGCorrData.corr()
HYGCorr

TotalReturnsHYG  HY_Index_OAS  HY_INDEX_PX  \
TotalReturnsHYG           1.000000     -0.811564     0.992894   
HY_Index_OAS             -0.811564      1.000000    -0.768256   
HY_INDEX_PX               0.992894     -0.768256     1.000000   
HY_INDEX_Spread          -0.823615      0.992935    -0.790237   
HY_INDEX_Z_Spread        -0.817676      0.996408    -0.780753   
10y_UST_INDEX_PX          0.257882      0.074523     0.359889   
IG_Index_OAS             -0.854725      0.962528    -0.816633   
IG_INDEX_Spread          -0.856428      0.959035    -0.818980   
IG_INDEX_PX               0.736774     -0.801273     0.661887   
IG_INDEX_Z_Spread        -0.852980      0.962433    -0.814627   
HYG_Volume               -0.065813      0.289790    -0.022540   
SPY_LAST_PX               0.935804     -0.660271     0.964375   
HYG_MOD_DUR_YAS          -0.849440      0.808441    -0.821753   

                   HY_INDEX_Spread  HY_INDEX_Z_Spread  10y_UST_INDEX_PX  \
TotalReturnsHYG          -0.823615          -0.817676          0.257882   
HY_Index_OAS              0.992935           0.996408          0.074523   
HY_INDEX_PX              -0.790237          -0.780753          0.359889   
HY_INDEX_Spread           1.000000           0.997182         -0.018951   
HY_INDEX_Z_Spread         0.997182           1.000000          0.015602   
10y_UST_INDEX_PX         -0.018951           0.015602          1.000000   
IG_Index_OAS              0.962233           0.963136          0.017747   
IG_INDEX_Spread           0.960126           0.960037          0.010547   
IG_INDEX_PX              -0.750315          -0.768292         -0.453386   
IG_INDEX_Z_Spread         0.962234           0.963204          0.020021   
HYG_Volume                0.263709           0.273261          0.288718   
SPY_LAST_PX              -0.697828          -0.681092          0.517664   
HYG_MOD_DUR_YAS           0.810395           0.812892         -0.050900   

                   IG_Index_OAS  IG_INDEX_Spread  IG_INDEX_PX  \
TotalReturnsHYG       -0.854725        -0.856428     0.736774   
HY_Index_OAS           0.962528         0.959035    -0.801273   
HY_INDEX_PX           -0.816633        -0.818980     0.661887   
HY_INDEX_Spread        0.962233         0.960126    -0.750315   
HY_INDEX_Z_Spread      0.963136         0.960037    -0.768292   
10y_UST_INDEX_PX       0.017747         0.010547    -0.453386   
IG_Index_OAS           1.000000         0.999020    -0.818949   
IG_INDEX_Spread        0.999020         1.000000    -0.816758   
IG_INDEX_PX           -0.818949        -0.816758     1.000000   
IG_INDEX_Z_Spread      0.999927         0.999084    -0.819040   
HYG_Volume             0.249320         0.241821    -0.259552   
SPY_LAST_PX           -0.702898        -0.703997     0.497732   
HYG_MOD_DUR_YAS        0.886538         0.893106    -0.774240   

                   IG_INDEX_Z_Spread  HYG_Volume  SPY_LAST_PX  HYG_MOD_DUR_YAS  
TotalReturnsHYG            -0.852980   -0.065813     0.935804        -0.849440  
HY_Index_OAS                0.962433    0.289790    -0.660271         0.808441  
HY_INDEX_PX                -0.814627   -0.022540     0.964375        -0.821753  
HY_INDEX_Spread             0.962234    0.263709    -0.697828         0.810395  
HY_INDEX_Z_Spread           0.963204    0.273261    -0.681092         0.812892  
10y_UST_INDEX_PX            0.020021    0.288718     0.517664        -0.050900  
IG_Index_OAS                0.999927    0.249320    -0.702898         0.886538  
IG_INDEX_Spread             0.999084    0.241821    -0.703997         0.893106  
IG_INDEX_PX                -0.819040   -0.259552     0.497732        -0.774240  
IG_INDEX_Z_Spread           1.000000    0.248514    -0.700207         0.887442  
HYG_Volume                  0.248514    1.000000     0.022793         0.189483  
SPY_LAST_PX                -0.700207    0.022793     1.000000        -0.697568  
HYG_MOD_DUR_YAS             0.887442    0.189483    -0.697568         1.000000

Following, I created lagged variables by one day of all the previous information to run the regressions.

In [158]:
HYGCorrDataLagVol = HYGCorrData.copy()
HYGCorrDataLagVol.loc[:,"HY_Index_OAS_Lag_1"]=HYGCorrDataLagVol.loc[:,"HY_Index_OAS"].shift(-1)
HYGCorrDataLagVol.loc[:,"HY_INDEX_PX_Lag_1"]=HYGCorrDataLagVol.loc[:,"HY_INDEX_PX"].shift(-1)
HYGCorrDataLagVol.loc[:,"HY_INDEX_Spread_Lag_1"]=HYGCorrDataLagVol.loc[:,"HY_INDEX_Spread"].shift(-1)
HYGCorrDataLagVol.loc[:,"HY_INDEX_Z_Spread_Lag_1"]=HYGCorrDataLagVol.loc[:,"HY_INDEX_Z_Spread"].shift(-1)
HYGCorrDataLagVol.loc[:,"10y_UST_INDEX_PX_Lag_1"]=HYGCorrDataLagVol.loc[:,"10y_UST_INDEX_PX"].shift(-1)
HYGCorrDataLagVol.loc[:,"IG_Index_OAS_Lag_1"]=HYGCorrDataLagVol.loc[:,"IG_Index_OAS"].shift(-1)
HYGCorrDataLagVol.loc[:,"IG_Index_Spread_Lag_1"]=HYGCorrDataLagVol.loc[:,"IG_INDEX_Spread"].shift(-1)
HYGCorrDataLagVol.loc[:,"IG_INDEX_PX_Lag_1"]=HYGCorrDataLagVol.loc[:,"IG_INDEX_PX"].shift(-1)
HYGCorrDataLagVol.loc[:,"IG_INDEX_Z_Spread_Lag_1"]=HYGCorrDataLagVol.loc[:,"IG_INDEX_Z_Spread"].shift(-1)
HYGCorrDataLagVol.loc[:,"HYG_Volume_Lag_1"]=HYGCorrDataLagVol.loc[:,"HYG_Volume"].shift(-1)
HYGCorrDataLagVol.loc[:,"SPY_LAST_PX_Lag_1"]=HYGCorrDataLagVol.loc[:,"SPY_LAST_PX"].shift(-1)
HYGCorrDataLagVol.loc[:,"HYG_MOD_DUR_YAS_Lag_1"]=HYGCorrDataLagVol.loc[:,"HYG_MOD_DUR_YAS"].shift(-1)
HYGCorrDataLagVol.loc[:,"HY_OASxIG_OAS_Lag_1"]=(HYGCorrDataLagVol.loc[:,"HY_Index_OAS_Lag_1"]*HYGCorrDataLagVol.loc[:,"IG_Index_OAS_Lag_1"])
HYGCorrDataLagVol=HYGCorrDataLagVol.dropna()
HYGCorrDataLagVol.tail()

TotalReturnsHYG  HY_Index_OAS  HY_INDEX_PX  HY_INDEX_Spread  \
Date                                                                      
2021-01-19      8744.551454      3.475033      2349.11       374.901344   
2021-01-15      8717.841195      3.496868      2346.26       379.560074   
2021-01-14      8730.130922      3.471442      2345.24       377.592235   
2021-01-13      8730.420672      3.541150      2340.98       384.975631   
2021-01-12      8702.710408      3.558607      2338.69       387.074278   

            HY_INDEX_Z_Spread  10y_UST_INDEX_PX  IG_Index_OAS  \
Date                                                            
2021-01-19           375.5320            1.0886      0.924479   
2021-01-15           378.4480            1.0835      0.937524   
2021-01-14           375.8447            1.1292      0.929280   
2021-01-13           383.3776            1.0832      0.934683   
2021-01-12           385.9547            1.1291      0.949886   

            IG_INDEX_Spread  IG_INDEX_PX  IG_INDEX_Z_Spread  ...  \
Date                                                         ...   
2021-01-19        91.843802      3527.74            92.5690  ...   
2021-01-15        92.796750      3519.97            93.8883  ...   
2021-01-14        92.057996      3514.54            93.0581  ...   
2021-01-13        92.790383      3522.23            93.5878  ...   
2021-01-12        94.189026      3500.83            95.1180  ...   

            HY_INDEX_Z_Spread_Lag_1  10y_UST_INDEX_PX_Lag_1  \
Date                                                          
2021-01-19                 378.4480                  1.0835   
2021-01-15                 375.8447                  1.1292   
2021-01-14                 383.3776                  1.0832   
2021-01-13                 385.9547                  1.1291   
2021-01-12                 381.0291                  1.1460   

            IG_Index_OAS_Lag_1  IG_Index_Spread_Lag_1  IG_INDEX_PX_Lag_1  \
Date                                                                       
2021-01-19            0.937524              92.796750            3519.97   
2021-01-15            0.929280              92.057996            3514.54   
2021-01-14            0.934683              92.790383            3522.23   
2021-01-13            0.949886              94.189026            3500.83   
2021-01-12            0.949604              94.272522            3501.58   

            IG_INDEX_Z_Spread_Lag_1  HYG_Volume_Lag_1  SPY_LAST_PX_Lag_1  \
Date                                                                       
2021-01-19                  93.8883        48134776.0         384.364921   
2021-01-15                  93.0581        25484782.0         382.854922   
2021-01-14                  93.5878        47710312.0         384.214922   
2021-01-13                  95.1180        41746060.0         383.864922   
2021-01-12                  95.0572        26733124.0         378.624921   

            HYG_MOD_DUR_YAS_Lag_1  HY_OASxIG_OAS_Lag_1  
Date                                                    
2021-01-19                 2.8757             3.278398  
2021-01-15                 2.8614             3.225942  
2021-01-14                 2.9157             3.309853  
2021-01-13                 2.9117             3.380271  
2021-01-12                 2.8659             3.347363  

[5 rows x 26 columns]

I tried several OLS regressions and all of them have a high multicolineariy indicatior of conditional number. For example look at the one below:

In [129]:
X = sm.add_constant(HYGCorrDataLagVol[["HY_Index_OAS","10y_UST_INDEX_PX","IG_Index_OAS","SPY_LAST_PX","HY_OASxIG_OAS"]])
Y = HYGCorrDataLagVol["TotalReturnsHYG"]/100

model = sm.OLS(Y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        TotalReturnsHYG   R-squared:                       0.969
Model:                            OLS   Adj. R-squared:                  0.969
Method:                 Least Squares   F-statistic:                     7715.
Date:                Sat, 10 Jan 2026   Prob (F-statistic):               0.00
Time:                        21:16:43   Log-Likelihood:                -1855.1
No. Observations:                1244   AIC:                             3722.
Df Residuals:                    1238   BIC:                             3753.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const               60.8299      1.222     49.792      0.000      58.433      63.227
HY_Index_OAS         3.6443      0.346     10.518      0.000       2.965       4.324
10y_UST_INDEX_PX    -1.0946      0.043    -25.537      0.000      -1.179      -1.011
IG_Index_OAS        -2.3016      0.909     -2.532      0.011      -4.085      -0.518
SPY_LAST_PX          0.0608      0.001     77.246      0.000       0.059       0.062
HY_OASxIG_OAS       -2.0017      0.224     -8.947      0.000      -2.441      -1.563
==============================================================================
Omnibus:                      190.357   Durbin-Watson:                   0.175
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              425.853
Skew:                           0.867   Prob(JB):                     3.37e-93
Kurtosis:                       5.283   Cond. No.                     2.43e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.43e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

For this reason, I am limited the OLS time series regressions to the Option adjusted spreads from UST for both High Yields and Investment grades bonds (including a combined dependent variable), and the 10 Year US Treasury yields. 

My logic says that the price and returns would depende on how much is paying over the UST, and if the Spread is high for investment grades & lower than usual for high yield then people would switch from one asset class to the other, which would reduce the price by demand. 

In [128]:
X = sm.add_constant(HYGCorrDataLagVol[["HY_Index_OAS","10y_UST_INDEX_PX","IG_Index_OAS","HY_OASxIG_OAS"]])
Y = HYGCorrDataLagVol["TotalReturnsHYG"]/100

model = sm.OLS(Y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        TotalReturnsHYG   R-squared:                       0.819
Model:                            OLS   Adj. R-squared:                  0.818
Method:                 Least Squares   F-statistic:                     1402.
Date:                Sat, 10 Jan 2026   Prob (F-statistic):               0.00
Time:                        21:16:41   Log-Likelihood:                -2950.6
No. Observations:                1244   AIC:                             5911.
Df Residuals:                    1239   BIC:                             5937.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const              127.3842      2.089     60.988      0.000     123.286     131.482
HY_Index_OAS        -6.9500      0.767     -9.058      0.000      -8.455      -5.445
10y_UST_INDEX_PX     1.4496      0.066     21.914      0.000       1.320       1.579
IG_Index_OAS       -35.3284      1.934    -18.265      0.000     -39.123     -31.534
HY_OASxIG_OAS        4.7888      0.496      9.652      0.000       3.815       5.762
==============================================================================
Omnibus:                       63.281   Durbin-Watson:                   0.032
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               71.730
Skew:                           0.581   Prob(JB):                     2.65e-16
Kurtosis:                       2.820   Cond. No.                         247.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Let's Try above expression lagged by one Day to see if that solves the structure


In [155]:
X = sm.add_constant(HYGCorrDataLagVol[["HY_Index_OAS_Lag_1","10y_UST_INDEX_PX_Lag_1","IG_Index_OAS_Lag_1","HY_OASxIG_OAS_Lag_1"]])
Y = HYGCorrDataLagVol["TotalReturnsHYG"]/100

model = sm.OLS(Y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        TotalReturnsHYG   R-squared:                       0.816
Model:                            OLS   Adj. R-squared:                  0.816
Method:                 Least Squares   F-statistic:                     1378.
Date:                Sat, 10 Jan 2026   Prob (F-statistic):               0.00
Time:                        22:18:37   Log-Likelihood:                -2959.4
No. Observations:                1244   AIC:                             5929.
Df Residuals:                    1239   BIC:                             5954.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                    127.6079      2.104     60.649      0.000     123.480     131.736
HY_Index_OAS_Lag_1        -6.8365      0.772     -8.859      0.000      -8.351      -5.323
10y_UST_INDEX_PX_Lag_1     1.4541      0.066     21.868      0.000       1.324       1.585
IG_Index_OAS_Lag_1       -36.2624      1.949    -18.608      0.000     -40.086     -32.439
HY_OASxIG_OAS_Lag_1        4.8806      0.500      9.769      0.000       3.900       5.861
==============================================================================
Omnibus:                       63.278   Durbin-Watson:                   0.045
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               72.070
Skew:                           0.585   Prob(JB):                     2.24e-16
Kurtosis:                       2.856   Cond. No.                         247.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The conditional number is still pretty high: 247, I decided to examine each variable one by one, and it increases really fast.

In [142]:
X = sm.add_constant(HYGCorrDataLagVol[["10y_UST_INDEX_PX"]])
Y = HYGCorrDataLagVol["TotalReturnsHYG"]/100

model = sm.OLS(Y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        TotalReturnsHYG   R-squared:                       0.066
Model:                            OLS   Adj. R-squared:                  0.066
Method:                 Least Squares   F-statistic:                     88.31
Date:                Sat, 10 Jan 2026   Prob (F-statistic):           2.59e-20
Time:                        21:21:24   Log-Likelihood:                -3971.2
No. Observations:                1244   AIC:                             7946.
Df Residuals:                    1242   BIC:                             7957.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const               84.7667      0.522    162.377      0.000      83.743      85.791
10y_UST_INDEX_PX     1.3760      0.146      9.397      0.000       1.089       1.663
==============================================================================
Omnibus:                      173.223   Durbin-Watson:                   0.005
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               43.267
Skew:                          -0.030   Prob(JB):                     4.02e-10
Kurtosis:                       2.088   Cond. No.                         11.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [143]:
X = sm.add_constant(HYGCorrDataLagVol[["10y_UST_INDEX_PX","HY_Index_OAS"]])
Y = HYGCorrDataLagVol["TotalReturnsHYG"]/100

model = sm.OLS(Y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        TotalReturnsHYG   R-squared:                       0.761
Model:                            OLS   Adj. R-squared:                  0.760
Method:                 Least Squares   F-statistic:                     1971.
Date:                Sat, 10 Jan 2026   Prob (F-statistic):               0.00
Time:                        21:21:51   Log-Likelihood:                -3124.7
No. Observations:                1244   AIC:                             6255.
Df Residuals:                    1241   BIC:                             6271.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const              108.0618      0.470    229.994      0.000     107.140     108.984
10y_UST_INDEX_PX     1.7095      0.074     22.980      0.000       1.564       1.855
HY_Index_OAS        -7.0129      0.117    -59.986      0.000      -7.242      -6.784
==============================================================================
Omnibus:                       71.358   Durbin-Watson:                   0.041
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               82.856
Skew:                           0.629   Prob(JB):                     1.02e-18
Kurtosis:                       2.867   Cond. No.                         28.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [ ]:
X = sm.add_constant(HYGCorrDataLagVol[["10y_UST_INDEX_PX","HY_Index_OAS","HYG_MOD_DUR_YAS"]])
Y = HYGCorrDataLagVol["TotalReturnsHYG"]/100

model = sm.OLS(Y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        TotalReturnsHYG   R-squared:                       0.835
Model:                            OLS   Adj. R-squared:                  0.835
Method:                 Least Squares   F-statistic:                     2097.
Date:                Sat, 10 Jan 2026   Prob (F-statistic):               0.00
Time:                        21:27:06   Log-Likelihood:                -2891.9
No. Observations:                1244   AIC:                             5792.
Df Residuals:                    1240   BIC:                             5812.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const              120.0495      0.638    188.144      0.000     118.798     121.301
10y_UST_INDEX_PX     1.4189      0.063     22.552      0.000       1.295       1.542
HY_Index_OAS        -3.7546      0.168    -22.334      0.000      -4.084      -3.425
HYG_MOD_DUR_YAS     -6.2006      0.261    -23.730      0.000      -6.713      -5.688
==============================================================================
Omnibus:                      126.259   Durbin-Watson:                   0.047
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               53.154
Skew:                           0.302   Prob(JB):                     2.87e-12
Kurtosis:                       2.188   Cond. No.                         59.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [156]:
X = sm.add_constant(HYGCorrDataLagVol[["10y_UST_INDEX_PX_Lag_1","HY_Index_OAS_Lag_1","HYG_MOD_DUR_YAS_Lag_1"]])
Y = HYGCorrDataLagVol["TotalReturnsHYG"]/100

model = sm.OLS(Y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        TotalReturnsHYG   R-squared:                       0.833
Model:                            OLS   Adj. R-squared:                  0.832
Method:                 Least Squares   F-statistic:                     2057.
Date:                Sat, 10 Jan 2026   Prob (F-statistic):               0.00
Time:                        22:19:15   Log-Likelihood:                -2901.9
No. Observations:                1244   AIC:                             5812.
Df Residuals:                    1240   BIC:                             5832.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                    120.1178      0.641    187.341      0.000     118.860     121.376
10y_UST_INDEX_PX_Lag_1     1.4316      0.063     22.626      0.000       1.307       1.556
HY_Index_OAS_Lag_1        -3.6656      0.169    -21.694      0.000      -3.997      -3.334
HYG_MOD_DUR_YAS_Lag_1     -6.3157      0.262    -24.061      0.000      -6.831      -5.801
==============================================================================
Omnibus:                       89.882   Durbin-Watson:                   0.056
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               47.373
Skew:                           0.314   Prob(JB):                     5.17e-11
Kurtosis:                       2.279   Cond. No.                         59.5
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Final regression I obtained so far**

Has a 288 conditional number, p-values that can not deny the null hypothesis and a R-Squared that imply the variables could explain an 81% of the dependent variable variance. Do you think the Conditional Number is a concern? Once I add SPY variables the conditional number goes from a little high to extremely concerning. 

In [160]:
X = sm.add_constant(HYGCorrDataLagVol[["10y_UST_INDEX_PX_Lag_1","HY_Index_OAS_Lag_1","IG_Index_OAS_Lag_1","HY_OASxIG_OAS_Lag_1","HYG_MOD_DUR_YAS_Lag_1"]])
Y = HYGCorrDataLagVol["TotalReturnsHYG"]/100

model = sm.OLS(Y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        TotalReturnsHYG   R-squared:                       0.843
Model:                            OLS   Adj. R-squared:                  0.842
Method:                 Least Squares   F-statistic:                     1328.
Date:                Sat, 10 Jan 2026   Prob (F-statistic):               0.00
Time:                        22:32:38   Log-Likelihood:                -2862.9
No. Observations:                1244   AIC:                             5738.
Df Residuals:                    1238   BIC:                             5769.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                    133.0660      1.984     67.063      0.000     129.173     136.959
10y_UST_INDEX_PX_Lag_1     1.3935      0.062     22.586      0.000       1.272       1.515
HY_Index_OAS_Lag_1        -7.2581      0.715    -10.152      0.000      -8.661      -5.855
IG_Index_OAS_Lag_1       -18.6860      2.177     -8.582      0.000     -22.958     -14.414
HY_OASxIG_OAS_Lag_1        3.7924      0.469      8.093      0.000       2.873       4.712
HYG_MOD_DUR_YAS_Lag_1     -4.9904      0.346    -14.416      0.000      -5.670      -4.311
==============================================================================
Omnibus:                       73.322   Durbin-Watson:                   0.052
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               72.793
Skew:                           0.546   Prob(JB):                     1.56e-16
Kurtosis:                       2.540   Cond. No.                         288.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [157]:
HYGCorrDataLagVol.corr()

TotalReturnsHYG  HY_Index_OAS  HY_INDEX_PX  \
TotalReturnsHYG                 1.000000     -0.811602     0.992904   
HY_Index_OAS                   -0.811602      1.000000    -0.768333   
HY_INDEX_PX                     0.992904     -0.768333     1.000000   
HY_INDEX_Spread                -0.823654      0.992935    -0.790319   
HY_INDEX_Z_Spread              -0.817705      0.996409    -0.780820   
10y_UST_INDEX_PX                0.257647      0.074729     0.359580   
IG_Index_OAS                   -0.854967      0.962608    -0.816977   
IG_INDEX_Spread                -0.856755      0.959168    -0.819425   
IG_INDEX_PX                     0.737637     -0.801785     0.662905   
IG_INDEX_Z_Spread              -0.853244      0.962526    -0.814996   
HYG_Volume                     -0.066016      0.289856    -0.022812   
SPY_LAST_PX                     0.936015     -0.660602     0.964508   
HYG_MOD_DUR_YAS                -0.850878      0.809331    -0.823427   
HY_Index_OAS_Lag_1             -0.807001      0.993049    -0.765912   
HY_INDEX_PX_Lag_1               0.991626     -0.767900     0.999519   
HY_INDEX_Spread_Lag_1          -0.819184      0.985740    -0.787931   
HY_INDEX_Z_Spread_Lag_1        -0.813044      0.989045    -0.778331   
10y_UST_INDEX_PX_Lag_1          0.259772      0.075354     0.360859   
IG_Index_OAS_Lag_1             -0.852253      0.958673    -0.815587   
IG_Index_Spread_Lag_1          -0.854200      0.955687    -0.818073   
IG_INDEX_PX_Lag_1               0.733751     -0.801771     0.660081   
IG_INDEX_Z_Spread_Lag_1        -0.850543      0.958584    -0.813592   
HYG_Volume_Lag_1               -0.063820      0.293562    -0.022359   
SPY_LAST_PX_Lag_1               0.936512     -0.662517     0.964905   
HYG_MOD_DUR_YAS_Lag_1          -0.849603      0.808907    -0.821661   
HY_OASxIG_OAS_Lag_1            -0.819667      0.983400    -0.779892   

                         HY_INDEX_Spread  HY_INDEX_Z_Spread  10y_UST_INDEX_PX  \
TotalReturnsHYG                -0.823654          -0.817705          0.257647   
HY_Index_OAS                    0.992935           0.996409          0.074729   
HY_INDEX_PX                    -0.790319          -0.780820          0.359580   
HY_INDEX_Spread                 1.000000           0.997182         -0.018892   
HY_INDEX_Z_Spread               0.997182           1.000000          0.015766   
10y_UST_INDEX_PX               -0.018892           0.015766          1.000000   
IG_Index_OAS                    0.962311           0.963227          0.017157   
IG_INDEX_Spread                 0.960257           0.960184          0.009741   
IG_INDEX_PX                    -0.750795          -0.768816         -0.452472   
IG_INDEX_Z_Spread               0.962325           0.963310          0.019378   
HYG_Volume                      0.263769           0.273339          0.288301   
SPY_LAST_PX                    -0.698183          -0.681407          0.516865   
HYG_MOD_DUR_YAS                 0.811283           0.813826         -0.053516   
HY_Index_OAS_Lag_1              0.985707           0.989278          0.076315   
HY_INDEX_PX_Lag_1              -0.789779          -0.780402          0.358773   
HY_INDEX_Spread_Lag_1           0.991828           0.989687         -0.016887   
HY_INDEX_Z_Spread_Lag_1         0.989474           0.992194          0.017436   
10y_UST_INDEX_PX_Lag_1         -0.018099           0.016792          0.998510   
IG_Index_OAS_Lag_1              0.957965           0.959171          0.020228   
IG_Index_Spread_Lag_1           0.956291           0.956564          0.013002   
IG_INDEX_PX_Lag_1              -0.750664          -0.769067         -0.454347   
IG_INDEX_Z_Spread_Lag_1         0.957964           0.959219          0.022533   
HYG_Volume_Lag_1                0.266602           0.277219          0.292162   
SPY_LAST_PX_Lag_1              -0.700183          -0.683429          0.516476   
HYG_MOD_DUR_YAS_Lag_1           0.810272           0.812975         -0.048349   
H